### ISSUE TITLE
Various issues iterating RealmResults

### ISSUE THREAD

**Issue by andriy-svirskyy** [realm/realm-java#7502]
#### Goal
App should not crash iterating RealmResults collection.

#### Actual Results
sometimes **invalid index exception** (length=15; index=15)

java.util.ArrayList.add(ArrayList.java:442); 
io.realm.internal.OsSharedRealm.addIterator(OsSharedRealm.java:495); 
io.realm.internal.OsResults$Iterator.<init>(OsResults.java:68); 
io.realm.OrderedRealmCollectionImpl$RealmCollectionIterator.<init>(OrderedRealmCollectionImpl.java:531); 
io.realm.OrderedRealmCollectionImpl.iterator(OrderedRealmCollectionImpl.java:221); 
io.realm.RealmResults.iterator(RealmResults.java:71); 

sometimes exception [**This Realm instance has already been closed, making it unusable.**]

io.realm.internal.OsResults$Iterator.<init>(OsResults.java:55); 
io.realm.OrderedRealmCollectionImpl$RealmCollectionIterator.<init>(OrderedRealmCollectionImpl.java:531); 
io.realm.OrderedRealmCollectionImpl.iterator(OrderedRealmCollectionImpl.java:221); 
io.realm.RealmResults.iterator(RealmResults.java:71); 

#### Steps & Code to Reproduce
Issue observed randomly at the same line of code without obvious pattern. Majority of times app performs correctly.

The loop is:

```
for(@NonNull final Outbox current: outbox) {
        @Nullable final Item item = current.item();
        if(null == item) continue;
        
        processed.addLast(current.uid());
        if(!item.isValid()) continue;
        
        output.addLast(item.uid());
}
```

Where `outbox` is

```
realm.where(Outbox.class)
        .lessThanOrEqualTo(Outbox.field__dwelling(), tm_threshold)
        .findAll()
        .freeze()
```

Before the loop log output confirms that `outbox.getRealm().isClosed()==false`. Crash happens at the `for()` line.

#### Version of Realm and tooling

Realm version(s): io.realm:realm-gradle-plugin:10.6.0

Realm Sync feature enabled: No

Android Studio version: 4.2.1

Android Build Tools version: 30.0.3

Gradle version: 3.3.1

Which Android version and device(s): Android 7.0, Huawei P9 EVA-L09

**Comment by cmelchior** [comment]
Hi @andriy-svirskyy. Thank you for the report. It is a bit hard to follow the logic in your example. It seems to call into some code not part of your sample, also, I'm not sure what the `freeze()` method does?

Is it possible to create a minimal sample case that reproduces the issue we can debug further into?

**Comment by andriy-svirskyy** [comment]
More detailed code sample:

```
try(@NonNull Realm realm = Realm.getDefaultInstance()) {
	@NonNull final RealmResults<Outbox> outbox = realm.where(Outbox.class).lessThanOrEqualTo(Outbox.field__dwelling(), tm_threshold).findAll().freeze();
	final ArrayDeque<Long> output = new ArrayDeque<>();
	final ArrayDeque<ObjectId> processed = new ArrayDeque<>();
	
	for(@NonNull final Outbox current: outbox) {
		@Nullable final Item item = current.item();
		if(null == item) continue;

		processed.addLast(current.uid());
		if(!item.isValid()) continue;

		output.addLast(item.uid());
	}
}
```

Exception is thrown from the line which reads `for(@NonNull final Outbox current: outbox) {`

Exception messages suggest that `outbox` gets changed despite of being frozen.

I use `freeze()` method to avoid `if(!current.isValid())` in the loop.

And more complete stack traces:

```
io.realm.internal.OsResults$Iterator.<init>(OsResults.java:55)
io.realm.OrderedRealmCollectionImpl$RealmCollectionIterator.<init>(OrderedRealmCollectionImpl.java:531)
io.realm.OrderedRealmCollectionImpl.iterator(OrderedRealmCollectionImpl.java:221)
io.realm.RealmResults.iterator(RealmResults.java:71)
sample(Instant.java:131)
android.os.Handler.handleCallback(Handler.java:761)
android.os.Handler.dispatchMessage(Handler.java:98)
android.os.Looper.loop(Looper.java:156)
android.os.HandlerThread.run(HandlerThread.java:61)
```

and

```
java.util.ArrayList.add(ArrayList.java:442)
io.realm.internal.OsSharedRealm.addIterator(OsSharedRealm.java:495)
io.realm.internal.OsResults$Iterator.<init>(OsResults.java:68)
io.realm.OrderedRealmCollectionImpl$RealmCollectionIterator.<init>(OrderedRealmCollectionImpl.java:531)
io.realm.OrderedRealmCollectionImpl.iterator(OrderedRealmCollectionImpl.java:221)
io.realm.RealmResults.iterator(RealmResults.java:71)
sample.update(Instant.java:131)
android.os.Handler.handleCallback(Handler.java:761)
android.os.Handler.dispatchMessage(Handler.java:98)
android.os.Looper.loop(Looper.java:156)
android.os.HandlerThread.run(HandlerThread.java:61)
```

I leave app running for 24 hours and get a couple of these crashes, while this procedure executed dozens of times.

**Comment by andriy-svirskyy** [comment]
I can confirm that applying `freeze` to every element in collection instead of query result resolves the issue. While this solution can be the workaround, the original issue stays unresolved.

The code now looks like this:

```
try(@NonNull Realm realm = Realm.getDefaultInstance()) {
	@NonNull final RealmResults<Outbox> outbox = realm.where(Outbox.class).lessThanOrEqualTo(Outbox.field__dwelling(), tm_threshold).findAll();// <-- modification
	final ArrayDeque<Long> output = new ArrayDeque<>();
	final ArrayDeque<ObjectId> processed = new ArrayDeque<>();
	
	for(@NonNull Outbox current: outbox) {
                current = current.freeze();// <-- modification
                if(!current.isValid()) continue;// <-- modification

		@Nullable final Item item = current.item();
		if(null == item) continue;

		processed.addLast(current.uid());
		if(!item.isValid()) continue;

		output.addLast(item.uid());
	}
}
```

**Comment by andriy-svirskyy** [comment]
> ➤ Clemente Tort Barbero commented:
> 
> You could try using `createSnapshot` instead of `freeze`? It would return a stable collection that you can iterate on.

I need collection of 'non-live' elements, which as far as I understand can't be achieved by creating the snapshot.

**Comment by rorbech** [comment]
Closing this as it seems answered. If more assistance is need or you think there is an actual issue, please add more comments.
